### PR TITLE
fix(sql): fix not equals filter on indexed symbol column

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Fixes the following problems for queries with a not equals filter over an indexed symbol column (e.g. `select * from tab where sym != 42`):
* The underlying record cursor factory was filtering out NULL values leading to different behavior than the same filter non-indexed column
* In case of a filter with bind variables, the factory didn't recalculate the included set properly on bind variable values change
* Each query execution led to `ObjList::indexOf` (array scan, O(N) operation) done on each key. Moreover, `CharSequence::toString` done for each symbol value led to some litter. To fix that `ObjList`s containing symbol values were replaced with `IntHashSet`s with symbol keys.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 32 / 32 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java|32|32|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 32 / 32 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java|32|32|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 32 / 32 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java|32|32|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 32 / 32 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java|32|32|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java [hunk 1]
```diff
@@ -94,39 +102,65 @@ public boolean recordCursorSupportsRandomAccess() {
...
+            if (value != null) {
+                excludedKeys.add(symbolMapReader.keyOf(value));
+            } else {
+                excludeNull = true;
             }
         }
+
+        // Append new keys to the included set filtering out the excluded ones.
+        for (int k = 0, n = symbolMapReader.getSymbolCount(); k < n; k++) {
+            if (!excludedKeys.contains(k) && !includedKeys.contains(k)) {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java:128]:
```suggestion
            if (!excludedKeys.contains(k) && !includedKeys.add(k)) {
```

and then remove next line

File: core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java [hunk 2]
```diff
@@ -94,39 +102,65 @@ public boolean recordCursorSupportsRandomAccess() {
...
+                includedKeys.add(k);
+                addRowCursorFactory(k);
+            }
+        }
+
+        if (!excludeNull && symbolMapReader.containsNullValue() && !includedKeys.contains(SymbolTable.VALUE_IS_NULL)) {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java:134]: same here, add returns boolean

File: core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java [hunk 3]
```diff
@@ -94,39 +102,65 @@ public boolean recordCursorSupportsRandomAccess() {
...
-                    );
-                }
-                includedValues.add(Chars.toString(symbol));
-                cursorFactories.add(rowCursorFactory);
+
+        boolean excludeNull = false;
+        // Generate excluded key set.
+        for (int i = 0, n = keyExcludedValueFunctions.size(); i < n; i++) {
+            final CharSequence value = keyExcludedValueFunctions.getQuick(i).getStr(null);
+            if (value != null) {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java:119]: I don't think this `if` is necessary. `keyOf` can work with nulls

**Review by ideoma — COMMENTED** [review]
Minor comments overall.
Generally I'm surprised ExcludedValues symbol factory exists, I doubt that fetching by index not equals can yield any benefit from brute force (except when not equals is a list with 90% of values)

File: core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java (hunk 3 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java]: Addressed in b1075cc62e628100e930c525cdc36b94294761de

File: core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java (hunk 2 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java]: Addressed in abdb5e857e3af489eb61cdba7877a49a4af26474

File: core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java (hunk 1 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java]: Addressed in abdb5e857e3af489eb61cdba7877a49a4af26474

File: core/src/test/java/io/questdb/cutlass/pgwire/PGJobContextTest.java [hunk 4]
```diff
@@ -4163,6 +4163,74 @@ public void testUtf8QueryText() throws Exception {
...
+                                        "70.94360487171201,PEHN,1970-01-04 11:20:00.0\n",
+                                sink,
+                                rs
+                        );
+                    }
+                }
+
+                // the factory should correctly recognize NULL as the excluded value
+                sink.clear();
+                try (PreparedStatement ps = connection.prepareStatement("select * from x where b != ?")) {
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/test/java/io/questdb/cutlass/pgwire/PGJobContextTest.java:4217]: Added a new test in 6358909c56858669fc3569113f1b014708c87ff7. Hopefully, it reproduces the scenario you've described.

**Comment by puzpuzpuz** [comment]
> I'd suggest to seriously consider removing this "optimisation" or disable it unless as above filtering leaves only 10% of symbols to scan by index

The intent of this PR is to fix what's been broken for a long time.

I'd personally prefer to remove this factory completely rather than rely on 90% excludes keys (or any other) rule. The proper way to make a decision on index-based excluded values is to do a cost analysis. The 10% remaining keys may be used in the majority of rows and in this case a full scan will significantly outperform this "optimization". So, we need to gather and then calculate the percentage of rows, not keys. With bind variables, it gets more tricky since the values may change between query executions. In this case, we could decide on this optimization only if all of the keys have equal cardinality (with such an execution plan we'll have to hope that this will not change in the future when new rows are inserted).

@ideoma @bluestreak01 WDYT?

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 28 / 28 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java|28|28|100.00%|

**Comment by bluestreak01** [comment]
If you were to remove this factory, how do you see `sum != 'x'` executing ?

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 28 / 28 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java|28|28|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 28 / 28 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java|28|28|100.00%|

**Comment by bziobrowski** [comment]
I reckon that sym != 'x' or sym not in ( 'x', 'y', 'z' )  should rather execute as FilterOnValuesRecordCursorFactory using remaining index values (if there aren't too many) or  just a full table scan (taking ternary null logic into account). Adding proper stats (cardinality, nulls, histograms) would help but could also complicate query planning (as Andrey pointed out) because we'd either have to assume static query plan or add pre-assumptions about bind variables cardinalities . For example in Oracle one query can be associated with many different execution plans generated for specific set of bind variable values .

**Comment by puzpuzpuz** [comment]
> I reckon that sym != 'x' or sym not in ( 'x', 'y', 'z' )  should rather execute as FilterOnValuesRecordCursorFactory using remaining index values (if there aren't too many)

It won't be simple to implement since FilterOnValuesRecordCursorFactory assumes a statically sized set of included values. A full filtered scan is a more straightforward option.

**Comment by puzpuzpuz** [comment]
> If you were to remove this factory, how do you see `sum != 'x'` executing ?

As mentioned in https://github.com/questdb/questdb/pull/1838#issuecomment-1024900315, I'd do it via a full scan with a filter (potentially, a JIT-compiled one). Later, if we decide to implement a cost-based query optimizer, we could revisit this (and not only) case.

**Comment by puzpuzpuz** [comment]
Pushed 8c7b38de38aaa1176bae54960d143f6b20eac581 to reuse `RowCursorFactory` instances between subsequent query executions. @ideoma PTAL.

As for the way forward, we should decide whether we want to keep the excluded indexed symbol values optimization. If we want to keep it, at least for now, let's review and merge this PR since it contains important bugfixes.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 33 / 35 (94.29%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/SymbolIndexFilteredRowCursorFactory.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java|31|31|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/SymbolIndexRowCursorFactory.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 33 / 35 (94.29%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/SymbolIndexFilteredRowCursorFactory.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java|31|31|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/SymbolIndexRowCursorFactory.java|2|2|100.00%|

**Review by ideoma — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 33 / 35 (94.29%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/SymbolIndexFilteredRowCursorFactory.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java|31|31|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/SymbolIndexRowCursorFactory.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 33 / 35 (94.29%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/SymbolIndexFilteredRowCursorFactory.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java|31|31|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/SymbolIndexRowCursorFactory.java|2|2|100.00%|

**Pull request merged**